### ISSUE TITLE
Messages can have a ton of trailing whitespace

### ISSUE THREAD

**Issue by holmesworcester** [TryQuiet/quiet#1618]
Steps to reproduce:

1. write a message with a ton of blank lines at the beginning and the end of the message using shift+enter
2. send it

expected: empty preceding and trailing lines are stripped and do not display, on the receiver side
actual: they all display, impacting the readability of the channel.

Note that we want to strip the empty lines on the receiver side, so that a malicious client can't send messages that create large blank areas.

As an example of the correct behavior, look at Slack

**Comment by sbazerque** [comment]
I'll give this issue a go!

**Comment by sbazerque** [comment]
Idea: add logic to 

```
async canAppend(entry: LogEntry<ChannelMessage>) {
```

in `MessageAccessController` and reject the `entry` if the message contains trailing / heading whitespace.

It'd be wise to strip such whitespace upon entering a message in all clients, otherwise message sending would just fail in these cases.

**Comment by sbazerque** [comment]
OK - I added the validation on canAppend, and that prevents the message from being sent to the network apparently (it stays grayed out forever).

So I think what'd be missing to create a proper pull request would be to remove the whitespace in

1) the place where the data is ingested in the DB
2) the UI components that capture the message initially

(or maybe both? to ensure that the UI and the DB stay in sync?)

**Comment by holmesworcester** [comment]
My questions:

1. Should we handle messages from older clients with trailing whitespace, e.g. by allowing older versions to have whitespace but converting it on read? Where should this happen? MessagesAccessController?
2.  shouldn't the frontend get informative error messages when it sends a malformed message to the backend? Maybe the pattern here is for the backend to reject invalid messages from the frontend and say why.
3. Where is the best place to put validation on message entry? Ideally it seems like it would live in the shared state-manager code so that we have the same validation code on desktop and mobile, but this also seems like it overcomplicates things.

Here are some other places where we might need this checking, to give us an idea of the shape of the problem and the right solution:
1. Message length limits https://github.com/TryQuiet/quiet/issues/805
2. Limit on number of file attachments
3. Sanitizing messages e.g. stripping <script> tags (https://github.com/TryQuiet/quiet/issues/744)
4. Handling malicious images https://github.com/TryQuiet/quiet/issues/743
5. Invalid data in user table or channel table (part of this issue https://github.com/TryQuiet/quiet/issues/109)

**Comment by sbazerque** [comment]
> Should we handle messages from older clients with trailing whitespace, e.g. by allowing older versions to have whitespace but converting it on read? Where should this happen? MessagesAccessController?

Yes, I forgot to mention the versioning part! I think we need to add an optional version to the Message schema (we've already located that, it's the the file that uses the joi package for validation), but we still need to find where the actual message is constructed so we can add a version there. Maybe we can define a global version for the entire schema, instead of doing it per-entity? And then check the whitespace in `canAppend` only if the version is present. Converting on read sounds resonable to me for older schema versions.

> shouldn't the frontend get informative error messages when it sends a malformed message to the backend? Maybe the pattern here is for the backend to reject invalid messages from the frontend and say why.

Yes, having the error bubble up to the UI sounds good, but I'd do that independently of this change. For this change, I'd also modify the UI so that it always strips the heading / trailing whitespace - so there should be no new errors.

> Where is the best place to put validation on message entry? Ideally it seems like it would live in the shared state-manager code so that we have the same validation code on desktop and mobile, but this also seems like it overcomplicates things.

Let's investigate that a bit, or maybe someone can provide context for us?

**Comment by holmesworcester** [comment]
What's the most conventional solution to this?

**Comment by sbazerque** [comment]
I'd say convention is 

 - Verifying in the back-end, and have the error bubble up with a coherent error message to the UI
 - But _also_ verify in the point of entry in the UI, before sending to the backend, if possible, for better UX.

But that's pretty much what you were suggesting I believe

**Comment by holmesworcester** [comment]
> I'd say convention is
> 
> * Verifying in the back-end, and have the error bubble up with a coherent error message to the UI
> * But _also_ verify in the point of entry in the UI, before sending to the backend, if possible, for better UX.
> 
> But that's pretty much what you were suggesting I believe

Yeah, this sounds right.

**Comment by holmesworcester** [comment]
@sbazerque we talked about this a little today and we think we should just focus on the presentation layer for this whitespace issue rather than thinking about it like a data verification issue.  

We can create another issue for adding versioning. 

Do you see any downsides to that?

**Comment by sbazerque** [comment]
Hi @holmesworcester! Do you mean using input validation / formatting on read just for the whitespace case, or for all the issues that badly shaped data could cause?

Conceptually, I think it boils down to your threat model: if you still want to defend from disruption / bad behavior caused by peers that've been accepted to the community, I think validating in the data/sync layer is the only viable option. If, OTOH, community members are trusted, UI validation should be OK. That said, the whitespace issue doesn't really look that mighty as an attack vector (you could also just send a really long mesasge, right?).

> We can also let people send with lots of leading and trailing whitespace and simply strip it out in the display of messages.

Yeah, throwing an error for excessive whitespace feels odd. Discord, for example, just strips it.

> One downside I see is that, if we have some special way of displaying very long or very vertical messages (e.g. a limited number of lines and "read more...") we'll have to create some kind of special case for determining whether a message is indeed very long or just very full of leading and trailing whitespace.

Mmm I'd do it as things leave the database, not on the UI, as a general rule. It seems like lots of little problems like the one you just described could crop up over time otherwise.

> We can create another issue for adding versioning.

Sounds good, I'd suggest seeing if there are any immediate use cases for versioning (planned schema changes or such) and reason from there?

**Comment by holmesworcester** [comment]
@kowalski any objections to stripping whitespace from messages as they leave the database?

@sbazerque we might be adding them directly to the frontend state on the sender's side so that they display instantly. That's one reason I can think of why having the database do it might not be the best solution in this case. But I'm not sure if we do it this way.

**Comment by kowalski** [comment]
I think stripping needs to happen in the presentation layer.
Messages are displayed to the sender even before they are inserted to the
database. This is because we want the app to be responsive, even if network
connection is.. subpar.
We don't want the sender to see a different state than everyone else. It's
essential that whatever transformation is applied it's applied consistently.

śr., 26 lip 2023, 06:14 użytkownik holmesworcester ***@***.***>
napisał:

**Comment by sbazerque** [comment]
My feeling is that considering other members of the community as not-fully-trusted is still not part of the architecture (at least not yet), so I think @kowalski's point is valid and is the most pragmatic way forward!

**Comment by sbazerque** [comment]
It'd be terrific if the UI could just lift the state directly from the local store, though (not waiting for stuff to go out the network or anything). I feel this is the trend in local-first software, in my experience it's fast enough to not present any noticeable delays.

**Comment by holmesworcester** [comment]
My memory is that we made this change to add the message directly to state because the local CRDT we are using would be slow to update in some cases, for example when the user is syncing lots of new messages. It's also something the user is very sensitive to, because otherwise it looks like their message was lost for a split second, so our tolerance for delay is low. Maybe with a more performant CRDT we could make this simplification.

**Comment by holmesworcester** [comment]
@maledorak this is an interesting and very comprehensive solution to the questions that come up if we were not just hiding trailing whitespace in the frontend: https://www.inkandswitch.com/cambria/

(for your curiosity and enjoyment!)

**Comment by siepra** [comment]
@maledorak Will you work on that? Should I move it back to backlog?

**Comment by holmesworcester** [comment]
Moving it back to the backlog.
